### PR TITLE
feat: implement basic auth for pactus shell

### PR DESCRIPTION
**Pull request by amirvalhalla**
## Description

implement basic auth for pactus shell

### CODE REVIEW

File: cmd/shell/cmd/root.go [hunk 1]
```diff
@@ -8,7 +8,7 @@ import (
 
 // rootCmd represents the base command when called without any subcommands.
 var rootCmd = &cobra.Command{
-	Use:   "ctl",
+	Use:   "shell",
 	Short: "Pactus control cli",
```
Comment (b00f) [inline cmd/shell/cmd/root.go:12]:
```suggestion
	Short: "Pactus shell",
```

File: cmd/shell/cmd/client.go [hunk 2]
```diff
@@ -11,6 +19,22 @@ const (
 )
 
 func init() {
+	client.RegisterFlagBinder(func(fs *pflag.FlagSet, namer naming.Namer) {
+		fs.StringVar(&Auth.Username, namer("auth-username"), Auth.Username, "username for authentication")
```
Comment (b00f) [inline cmd/shell/cmd/client.go:22]: Default is "",

File: cmd/shell/cmd/client.go [hunk 3]
```diff
@@ -11,6 +19,22 @@ const (
...
+		fs.StringVar(&Auth.Password, namer("auth-password"), Auth.Password, "password for authentication")
+	})
+
+	client.RegisterPreDialer(func(_ context.Context, opts *[]grpc.DialOption) error {
+		cred := Auth
+
+		authHeader := fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(cred.Username+":"+cred.Password)))
```
Comment (b00f) [inline cmd/shell/cmd/client.go]: How about when gRPC doesn't have password?

File: cmd/shell/cmd/root.go [hunk 4]
```diff
@@ -8,7 +8,7 @@ import (
...
 	Long:  `Managing the Pactus node using CLI.`,
```
Comment (b00f) [inline cmd/shell/cmd/root.go:13]:
```suggestion
	Long:  `pactus-shell is a command line tool for interacting with the Pactus blockchain using gRPC.`,
```

File: cmd/shell/cmd/client.go (hunk 3 above)
Comment (b00f) [reply cmd/shell/cmd/client.go]: Please test it, with or without auth.

File: cmd/shell/cmd/client.go (hunk 2 above)
Comment (b00f) [reply cmd/shell/cmd/client.go:22]: rename it to `username` and `password`.
Help string should be like this: "username for gRPC basic authentication"

File: cmd/shell/cmd/client.go [hunk 5]
```diff
@@ -11,6 +19,22 @@ const (
 )
 
 func init() {
+	client.RegisterFlagBinder(func(fs *pflag.FlagSet, namer naming.Namer) {
+		fs.StringVar(&Auth.Username, namer("auth-username"), Auth.Username, "username for authentication")
+		fs.StringVar(&Auth.Password, namer("auth-password"), Auth.Password, "password for authentication")
```
Comment (b00f) [inline cmd/shell/cmd/client.go:23]: Same here

File: cmd/shell/cmd/auth.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+var Auth = &auth{}
+
+type auth struct {
+	Username string
+	Password string
+}
+
+// BasicAuthCredentials is a custom type implementing grpc.PerRPCCredentials.
+type basicAuthCredentials struct {
```
Comment (b00f) [inline cmd/shell/cmd/auth.go:13]: Explain why we need this struct?

**Review by b00f — CHANGES_REQUESTED** [review]

File: cmd/shell/cmd/client.go (hunk 3 above)
Comment (amirvalhalla) [reply cmd/shell/cmd/client.go]: it works fine.

File: util/auth.go [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package util
+
+import (
+	"encoding/base64"
+	"fmt"
+)
+
+func BasicAuth(username, password string) string {
```
Comment (b00f) [inline util/auth.go:8]: 1- Use more semantic name, like: `MakeBasicCredential`
Read more here: https://datatracker.ietf.org/doc/html/rfc7617

2- Mode it to `basicauth` package. Read above

File: cmd/shell/cmd/auth.go (hunk 6 above)
Comment (b00f) [reply cmd/shell/cmd/auth.go:13]: Move this to `grpc/basicauth` package.
Make sure http module also use this package.

File: util/auth_test.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+	"testing"
+)
+
+func TestBasicAuth(t *testing.T) {
+	testCases := []struct {
+		username string
+		password string
+		expected string
+	}{
+		{"user", "pass", "Basic dXNlcjpwYXNz"},
```
Comment (b00f) [inline util/auth_test.go:14]: Add this case from  rfc7617

123:123 => dGVzdDoxMjPCow==

File: cmd/shell/cmd/client.go (hunk 2 above)
Comment (b00f) [reply cmd/shell/cmd/client.go]: Please don't resolve the conversation without explaining why it should get closed.

**Review by b00f — CHANGES_REQUESTED** [review]

File: cmd/shell/cmd/client.go (hunk 5 above)
Comment (b00f) [reply cmd/shell/cmd/client.go]: Please don't close the conversation, without explaining why.

File: cmd/shell/cmd/client.go [hunk 9]
```diff
@@ -11,6 +18,20 @@ const (
 )
 
 func init() {
+	client.RegisterFlagBinder(func(fs *pflag.FlagSet, namer naming.Namer) {
+		fs.StringVar(&Auth.Username, namer("username"), Auth.Username, "username for gRPC basic authentication")
```
[the remark this answers is not in the document]
Comment (amirvalhalla) [reply cmd/shell/cmd/client.go:22]: yes, it's empty by default

[the remark this answers is not in the document]
Comment (b00f) [reply cmd/shell/cmd/client.go:24]: So please use "" to emphasize it is empty string. We do the same in other commands

**Comment by b00f** [comment]
@amirvalhalla LGTM. Thanks. Just one comment to merge

**Comment by amirvalhalla** [comment]
> @amirvalhalla LGTM. Thanks. Just one comment to merge

done

File: cmd/shell/cmd/client.go [hunk 10]
```diff
@@ -1,16 +1,39 @@
...
 )
 
+var Auth = &basicauth.BasicAuth{}
+
 const (
 	defaultServerAddr     = "localhost:50051"
 	defaultResponseFormat = "prettyjson"
 )
 
 func init() {
```
Comment (b00f) [inline cmd/shell/cmd/client.go:22]: One question, Can we add this code inside the main function and remove `init`?
Did you try this?

File: cmd/shell/cmd/client.go [hunk 11]
```diff
@@ -1,16 +1,39 @@
...
+	"github.com/NathanBaulch/protoc-gen-cobra/client"
+	"github.com/NathanBaulch/protoc-gen-cobra/naming"
+	"github.com/pactus-project/pactus/www/grpc/basicauth"
 	pb "github.com/pactus-project/pactus/www/grpc/gen/go"
 	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"google.golang.org/grpc"
 )
 
+var Auth = &basicauth.BasicAuth{}
```
Comment (b00f) [inline cmd/shell/cmd/client.go:15]: Still I am not convinced we need it.

File: www/grpc/basicauth/basicauth.go [hunk 12]
```diff
@@ -0,0 +1,27 @@
+package basicauth
+
+import (
+	"context"
+)
+
+type BasicAuth struct {
```
Comment (b00f) [inline www/grpc/basicauth/basicauth.go:7]: The file name and this struct are not matched. remove it.

File: www/grpc/basicauth/credential.go [hunk 13]
```diff
@@ -0,0 +1,13 @@
+package basicauth
```
Comment (b00f) [inline www/grpc/basicauth/credential.go:1]: Merge this file with above file

**Review by b00f — CHANGES_REQUESTED** [review]

**Review by b00f — APPROVED** [review]

**Pull request merged**